src/modules/go.rs

```diff
@@ -0,0 +1,93 @@
+use ansi_term::Color;
+use std::path::PathBuf;
+use std::process::Command;
+
+use super::{Context, Module};
+
+/// Creates a segment with the current Python version
+///
+/// Will display the Python version if any of the following criteria are met:
+///     - Current directory contains a `.go` file
+///     - Current directory contains a `go.mod` file
+///     - Current directory contains a `go.sum` file
+///     - Current directory contains a `Godeps` directory
+///     - Current directory contains a `glide.yaml` file
+///     - Current directory contains a `Gopkg.yml` file
+///     - Current directory contains a `Gopkg.lock` file
+pub fn segment(context: &Context) -> Option<Module> {
+    let is_go_project = context.dir_files.iter().any(has_go_files);
+    if !is_go_project {
+        return None;
+    }
+
+    match get_go_version() {
+        Some(go_version) => {
+            const GO_CHAR: &str = "🐹 ";
+            let module_color = Color::Cyan.bold();
+
+            let mut module = Module::new("go");
+            module.set_style(module_color);
+
+            let formatted_version = format_go_version(go_version);
+            module.new_segment("symbol", GO_CHAR);
+            module.new_segment("version", formatted_version);
+
+            Some(module)
+        }
+        None => None,
+    }
+}
+
+fn has_go_files(dir_entry: &PathBuf) -> bool {
+    let is_go_mod =
+        |d: &PathBuf| -> bool { d.is_file() && d.file_name().unwrap_or_default() == "go.mod" };
+    let is_go_sum =
+        |d: &PathBuf| -> bool { d.is_file() && d.file_name().unwrap_or_default() == "go.sum" };
+    let is_godeps =
+        |d: &PathBuf| -> bool { d.is_dir() && d.file_name().unwrap_or_default() == "Godeps" };
+    let is_glide_yaml =
+        |d: &PathBuf| -> bool { d.is_file() && d.file_name().unwrap_or_default() == "glide.yaml" };
+    let is_go_file =
+        |d: &PathBuf| -> bool { d.is_file() && d.extension().unwrap_or_default() == "go" };
+    let is_gopkg_yml =
+        |d: &PathBuf| -> bool { d.is_file() && d.file_name().unwrap_or_default() == "Gopkg.yml" };
+    let is_gopkg_lock =
+        |d: &PathBuf| -> bool { d.is_file() && d.file_name().unwrap_or_default() == "Gopkg.lock" };
+
+    is_go_mod(&dir_entry)
+        || is_go_sum(&dir_entry)
+        || is_godeps(&dir_entry)
+        || is_glide_yaml(&dir_entry)
+        || is_go_file(&dir_entry)
+        || is_gopkg_yml(&dir_entry)
+        || is_gopkg_lock(&dir_entry)
+}
+
+fn get_go_version() -> Option<String> {
+    match Command::new("go").arg("version").output() {
+        Ok(output) => Some(String::from_utf8(output.stdout).unwrap()),
+        Err(_) => None,
+    }
+}
+
+fn format_go_version(go_stdout: String) -> String {
```

Comment:
Disclaimer: The comment from @Multimo is what had me stop and offer help here. Plus it was a fun distraction.

---

Let's call the `format_go_version` currently in the patch, `format_go_version_1`. Running it through `cargo bench` on my box produces: 177 ns.

---

First let's remove an unnecessary allocation. Instead of creating a temporary `String` at line 77 after `trim()` we can leave it as a slice until the end:

```rust
let mut trimmed_version = go_stdout
    .trim_start_matches("go version go")
    .trim();

let offset = trimmed_version.find(' ').unwrap();
let formatted_version: String = trimmed_version[..offset].into();
```

This cleans this up a bit and improves performance to 131 ns.

---

Next, let's remove the panic by assuming if we can't find `' '` the string cuts off there.

```rust
let offset = trimmed_version.find(' ').unwrap_or_else(|| trimmed_version.len());
```

This decreases the performance to 133 ns but panics are bad.

---

Next let's try changing this up a bit with `split` 

```rust
go_stdout
    // split into ["", "1.12.4 linux/amd64"]
    .splitn(2, "go version go")
    // return "1.12.4 linux/amd64"
    .nth(1)?
    // split into ["1.12.4", "linux/amd64"]
    .split_whitespace()
    // return "1.12.4"
    .next()
    // format as "v1.12.4"
    .map(|version| format!("v{}", version))
```

This brings us down to **105 ns**. You'll notice that now we are also accounting for the failure case of a `go version` returning garbage output. This function would return `None` in that case.

---

We can still squeeze a bit more performance out of this. `format!` is currently relatively slow in Rust for such simple tasks as string concatenation.

```rust
// Use with_capacity to pre-allocate the string so we can guarantee there is only 1 allocation
let mut buf = String::with_capacity(version.len() + 1);
buf.push('v');
buf.push_str(version);
buf
```

This increases the performance to **77 ns** or **2.3x** faster than what we started. 

That was a fun exercise.